celery_batches/__init__.py

```diff
@@ -34,11 +35,21 @@ def consume_queue(queue):
 
     """
     get = queue.get_nowait
+    requests_to_requeue = []
```

Comment:
I suspect that handling of the `eta` would be better to position in the `Strategy` (which seems to be where it is in celery itself, see https://github.com/celery/celery/blob/fbda0089f08d7f2a8f00925dbc0b6e10bd779251/celery/worker/strategy.py#L179-L185

But maybe that's not possible?

Reply:
In `celery`, it appears the Strategy's `task_message_handler()` is responsible for scheduling the ETA request by using the `call_at()` function, which exists in some sort of timer object. I assume this timer object keeps track of the request and handles it at the appropriate time.

For `celery-batches` to do the same thing in its Strategy, it would need some similar method of tracking the request until the ETA time has passed. It seems the only way `celery-batches` can track requests over time is via its Queue. I could move the ETA logic to the Strategy, but I'm not sure what I would do with the request if its ETA time hasn't passed yet and it needs to be stored somewhere that isn't the Queue (since the Queue would only be for requests that are ready).

I think it makes sense to keep the ETA logic in `consume_queue()` and use the Queue as both temporary storage for requests that aren't ready, and temporary storage for requests that are ready. But I'm fairly new to both Celery and celery-batches, so I'm open to changing it if you think it makes sense.

Reply:
I think storing them in the `Queue` is fine, I guess I'm trying to avoid adding Celery-specific logic into `consume_queue`, which is a rather dumb helper right now.

I think the `Strategy` could implement the same thing, pretty much:

1. Consume the queue.
2. For each `Request`, check it is ready, if not put it back in the queue.
3. Call the batch task.

Either implementation pretty much has the same logic, I think? There is a bad failure mode either way though, if you have more than `flush_every` in your queue that have not yet reached their ETA than batches would just keep trying to fire the tasks, but actually have nothing to do.

Maybe it makes sense to do something more like...

1. Consume the queue + the pending requests queue.
2. For each `Request`, check it is ready, if not put in a pending requests queue.
3. Call the batch task (if there are any requests left).

That still has some oddities, but might be more OK? This means that a task with an ETA on it would fire within `flush_interval` of the ETA -- we should probably document how ETA + flushing works regardless.

What do you think? Is that reasonable to your application?

Reply:
I've pushed a commit that I think makes sense, and all unit tests passed for me. I've also manually tested my project that uses my celery-batches repo, and everything seems good there too.